include/n64sys.h

```
@@ -82,16 +82,105 @@
  */
 #define MEMORY_BARRIER() asm volatile ("" : : : "memory")
 
+/**
+ * @brief Number of updates to the count register per second
+ *
+ * The count register updates at "half maximum instruction issue rate".
+ * This appears to be half the CPU clock.  Every second, this many counts
+ * will have passed in the count register
+ */
+#define COUNTS_PER_SECOND (93750000/2)
+
+// Returns the signed difference of time between "from" and "to". If "from"
+// is before "to", the distance in time is positive, otherwise it is
+// negative .
+#define TICKS_DISTANCE(from, to) ((int32_t)((uint32_t)(to) - (uint32_t)(from)))
+
+// Returns true if "t1" is before "t2". This is similar to t1 < t2,
+// but it correctly handles timer overflows which are very frequent.
+//
+// Notice that the N64 counter overflows every 90 seconds, so it's
+// not possible to compare times that are more than 45 seconds apart.
+#define TICKS_BEFORE(t1, t2) ({ TICKS_DISTANCE(t1, t2) > 0; })
+
+// Reads the COP0 register $9 (count) into the provided variable.
+// It will always produce a 32 bit unsigned value.
+#define READ_TICKS(variable) do { \
+    _Static_assert(sizeof(variable) >= 4, "target variable is narrower than 32 bits"); \
```

Maybe you needed this because of the way the macro works now, but in general I wouldn't do this. Anywhere there is a function returning a variable, it gets silently truncated if it's assigned to another one of smaller type. This is just the way C works, I don't think there's a good reason for this specific macro to forbid this behavior.